hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java

```diff
@@ -121,16 +125,53 @@ private static MapSampleMetadata resolvePortable(
         fields.put(topName, new MapTableField(topName, QueryDataType.OBJECT, !fields.isEmpty(), topPath));
 
         return new MapSampleMetadata(
-            GenericQueryTargetDescriptor.DEFAULT,
-            jetMapMetadataResolver.resolvePortable(classDef, isKey),
-            new LinkedHashMap<>(fields)
+                GenericQueryTargetDescriptor.DEFAULT,
+                jetMapMetadataResolver.resolvePortable(classDef, isKey),
+                new LinkedHashMap<>(fields)
+        );
+    }
+
+    /**
+     * Resolve metadata from a portable object.
+     *
+     * @param schema Compact schema (a.k.a class definition)
+     * @param isKey  Whether this is a key.
+     * @return Metadata.
+     */
+    private static MapSampleMetadata resolveCompact(
+            @Nonnull Schema schema,
+            boolean isKey,
+            JetMapMetadataResolver jetMapMetadataResolver
+    ) {
+        LinkedHashMap<String, TableField> fields = new LinkedHashMap<>();
+
+        Map<String, QueryDataType> simpleFields = FieldsUtil.resolveCompact(schema);
+
+        for (Entry<String, QueryDataType> fieldEntry : simpleFields.entrySet()) {
+            String name = fieldEntry.getKey();
+            TableField oldValue = fields.put(name,
+                    new MapTableField(name, fieldEntry.getValue(), false, new QueryPath(name, isKey)));
+            assert oldValue == null;
+        }
+
+        // Add top-level object.
+        String topName = isKey ? QueryPath.KEY : QueryPath.VALUE;
+        QueryPath topPath = isKey ? QueryPath.KEY_PATH : QueryPath.VALUE_PATH;
+        // explicitly remove to have the newly-inserted topName at the end
+        fields.remove(topName);
+        fields.put(topName, new MapTableField(topName, QueryDataType.OBJECT, !fields.isEmpty(), topPath));
+
+        return new MapSampleMetadata(
+                GenericQueryTargetDescriptor.DEFAULT,
+                jetMapMetadataResolver.resolveCompact(schema, isKey),
+                new LinkedHashMap<>(fields)
```

Comment:
Why don't we directly return `fields`?

Reply:
I copied directly from Portable but `fields` is not shared with anyone else. I think we can use it directly. I will change portable as well.